File: examples/Max.py
#!/usr/bin/env python

# ------
# Max.py
# ------

# run-time polymorphism

class A (object) :
    def __init__ (self, i) :
        self.i = i

    def __lt__ (self, other) :
        return self.i < other.i

class B (A) :
    pass

def my_max (x, y, *bf) :
    if (bf is ()) :
        if x < y :
            return y
        return x
    elif bf[0](x, y) :
        return y
    else :
        return x

print "Max.py"

assert hasattr(int,   "__lt__")
assert hasattr(float, "__lt__")
assert hasattr(str,   "__lt__")
assert hasattr(A,     "__lt__")
assert hasattr(B,     "__lt__")

assert my_max(2,   3)   == 3
assert my_max(2.3, 4)   == 4
assert my_max(2,   4.5) == 4.5
assert my_max(2.3, 4.5) == 4.5

assert my_max("abc", "def") == "def"

x = A(2)
y = A(3)
assert my_max(x, y)                      is y
assert my_max(x, y, lambda a, b : b < a) is x
assert    max(x, y)                      is y

x = B(2)
y = B(3)
assert my_max(x, y)                      is y
assert my_max(x, y, lambda a, b : b < a) is x
assert    max(x, y)                      is y

print "Done."

File: examples/Max.hs
-- ------
-- Max.hs
-- ------

-- compile-time polymorphism

{-
assert   :: Bool   -> a -> a
putStrLn :: String      -> IO ()
return   :: a           -> IO a
-}

{-
class Eq a where
    (==) :: a -> a -> Bool

class (Eq a) => Ord a where
    (<) :: a -> a -> Bool

instance Eq A where
    A x == A y  =  x == y

instance Ord A where
    A x <  A y  =  x <  y
-}

import Control.Exception (assert)

data A =
    A Int
    deriving (Eq, Ord)

data B =
    B Int

instance Eq B where
    B x == B y  =  x == y

instance Ord B where
    B x <  B y  =  x <  y

data C =
    C Int

myMax1 :: (Ord a) => a -> a -> a
myMax1 x y =
    if x < y then
        y
    else
        x

myMax2 :: a -> a -> (a -> a-> Bool) -> a
myMax2 x y bf =
    if (bf x y) then
        y
    else
        x

main :: IO ()
main = do
    putStrLn "Max.hs"

    assert ((myMax1 (2::Int) 3) == 3) return ()
    assert ((  max  (2::Int) 3) == 3) return ()

    assert ((myMax1 (2.3::Float) 4) == 4) return ()
    assert ((  max  (2.3::Float) 4) == 4) return ()

    assert ((myMax1 "abc" "def") == "def") return ()
    assert ((  max  "abc" "def") == "def") return ()

    assert ((myMax1 (A 2) (A 3)) == (A 3)) return ()
    assert ((  max  (A 2) (A 3)) == (A 3)) return ()
    assert ((myMax1 (B 2) (B 3)) == (B 3)) return ()
--  assert ((  max  (B 2) (B 3)) == (B 3)) return ()
--  assert ((myMax1 (C 2) (C 3)) == (C 3)) return () -- No instance for (Ord C), (Eq C)
--  assert ((  max  (C 2) (C 3)) == (C 3)) return () -- No instance for (Ord C), (Eq C)

    assert ((myMax2 (A 2) (A 3) (\x y -> y < x)) == (A 2)) return ()
    assert ((myMax2 (B 2) (B 3) (\x y -> y < x)) == (B 2)) return ()
--  assert ((myMax2 (C 2) (C 3) (\x y -> y < x)) == (C 2)) return () -- -- No instance for (Ord C), (Eq C)

    putStrLn "Done."
